aten/src/ATen/native/cuda/Sort.cu

```diff
@@ -0,0 +1,231 @@
+    } else {
+      values.resize_as_(self);
+      values.copy_(self);
+    }
+    if (!indices.defined()) {
+      indices = at::zeros({}, self.options().dtype(kLong));
+    } else {
+      indices.resize_as_(self);
+      indices.zero_();
+    }
+    return {values, indices};
+  }
+
+  Tensor self_;
+  if (is_non_overlapping_and_dense && self.stride(dim) == 1) {
+    self_ = self;
+  } else {
+    auto new_strides_unsort = infer_dense_strides_dim_last(self, dim);
+    self_ = at::empty_strided(self.sizes(), new_strides_unsort, self.options());
+    self_.copy_(self);
+  }
+
+  Tensor values_tmp, indices_tmp;
+  void *values_ptr_;
+  int64_t *indices_ptr;
+  if (!values.defined()) {
+    if (is_non_overlapping_and_dense) {
+      values = at::empty_strided(self.sizes(), self.strides(), self.options());
+    } else {
+      auto strides = at::infer_dense_strides(self.sizes(), self.strides());
+      values = at::empty_strided(self.sizes(), strides, self.options());
+    }
+  } else {
+    TORCH_CHECK(self_.scalar_type() == values.scalar_type(),
+      "Unexpected dtype for values, expect ", self_.scalar_type(), ", got ", values.scalar_type());
+    values.resize_as_(self);
+  }
+  if (values.strides() != self_.strides()) {
+    values_tmp = at::empty_strided(self_.sizes(), self_.strides(), self_.options());
+    values_ptr_ = values_tmp.data_ptr();
+  } else {
+    values_ptr_ = values.data_ptr();
+  }
+
+  if (!indices.defined()) {
+    if (is_non_overlapping_and_dense) {
+      indices = at::empty_strided(self.sizes(), self.strides(), self.options().dtype(kLong));
+    } else {
+      auto strides = at::infer_dense_strides(self.sizes(), self.strides());
+      indices = at::empty_strided(self.sizes(), strides, self.options().dtype(kLong));
+    }
+  } else {
+    TORCH_CHECK(kLong == indices.scalar_type(),
+      "Unexpected dtype for values, expect torch.long, got ", indices.scalar_type());
+    indices.resize_as_(self);
+  }
+  if (indices.strides() != self_.strides()) {
+    indices_tmp = at::empty_strided(self_.sizes(), self_.strides(), self_.options().dtype(kLong));
+    indices_ptr = indices_tmp.data_ptr<int64_t>();
+  } else {
+    indices_ptr = indices.data_ptr<int64_t>();
+  }
+
+  if (numel == 0) {
+    return {values, indices};
+  }
+
+  int64_t numel_or_intmax = std::min(numel, static_cast<int64_t>(std::numeric_limits<int>::max()));
+  int64_t nbatch = (numel_or_intmax / nsort) * nsort;
+
+  AT_DISPATCH_ALL_TYPES_AND(kHalf, self_.scalar_type(), "sort", [&]{
+    const scalar_t *self_ptr = self_.data_ptr<scalar_t>();
+    auto values_ptr = reinterpret_cast<scalar_t *>(values_ptr_);
+    int64_t remaining = numel;
+    while (remaining > 0) {
+      int64_t n = std::min(remaining, nbatch);
+      int64_t nsegments = n / nsort;
+      int64_t segment_bits = std::max<int64_t>(1L, static_cast<int64_t>(std::ceil(std::log2(nsegments))));
+
+      auto indices_and_segment = at::empty({nsegments, nsort, 2}, indices.options());
```

Comment:
why do you need a long here? values are guaranteed to fit in `int`.

Reply:
Yes! And taking advantage that, I can get another 1.8x speedup!

Reply:
New benchmark updated in https://github.com/pytorch/pytorch/pull/54626#issue-600072099